drivers/media/i2c/imx477.c

```diff
@@ -1285,20 +1292,49 @@ static void imx477_adjust_exposure_range(struct imx477 *imx477,
 
 	/* Honour the VBLANK limits when setting exposure. */
 	exposure_max = imx477->mode->height + imx477->vblank->val -
-						IMX477_EXPOSURE_OFFSET;
+		       (IMX477_EXPOSURE_OFFSET << imx477->long_exp_shift);
 	exposure_def = min(exposure_max, imx477->exposure->val);
 	__v4l2_ctrl_modify_range(imx477->exposure, imx477->exposure->minimum,
 				 exposure_max, imx477->exposure->step,
 				 exposure_def);
 }
 
+static int imx477_set_frame_length(struct imx477 *imx477, unsigned int val)
+{
+	int ret = 0;
+
+	imx477->long_exp_shift = 0;
+
+	while (val > IMX477_FRAME_LENGTH_MAX) {
+		if (++imx477->long_exp_shift > IMX477_LONG_EXP_SHIFT_MAX) {
+			imx477->long_exp_shift = IMX477_LONG_EXP_SHIFT_MAX;
```

Comment:
And a second question, doesn't this mean that the range on VBLANK has been set wrongly if we exceed the limit? Or are there rounding errors?

Reply:
Yes, strictly speaking, we cannot go over SHIFT_MAX based on the VBLANK limits set.  I was just being a bit defensive with this extra check.  I'll remove it.